core/src/graph/number.rs

```diff
@@ -1,51 +1,291 @@
 use super::prelude::*;
 
-use rand::distributions::uniform::SampleUniform;
+use rand::distributions::uniform::SampleRange;
 
-use num::{CheckedAdd, One, Zero};
+use num::{One, Zero, CheckedAdd};
 
-use std::ops::{Add, Rem, Sub};
+use std::ops::{Bound, RangeBounds};
 
-pub struct UniformRangeStep<N>(RangeStep<N>);
+/// A custom extension of all the range types that uses
+/// [`Bound`](std::ops::Bound) on the left and on the right.
+#[derive(Clone)]
+struct AnyRange<N> {
+    low: Bound<N>,
+    high: Bound<N>
+}
 
-impl<N: PartialOrd + Zero + Display> TryFrom<RangeStep<N>> for UniformRangeStep<N> {
-    type Error = anyhow::Error;
+macro_rules! any_range_int_impl {
+    { $target:ty } => {
+        impl AnyRange<$target> {
+            fn is_empty(&self) -> bool {
+                match (&self.low, &self.high) {
+                    (Bound::Excluded(low), Bound::Included(high))
+                    | (Bound::Included(low), Bound::Excluded(high)) => low >= high,
+                    (Bound::Included(low), Bound::Included(high)) => low > high,
+                    (Bound::Excluded(low), Bound::Excluded(high)) => *low + 1 >= *high,
+                    _ => false
+                }
+            }
+        }
+
+        impl SampleRange<$target> for AnyRange<$target> {
+            fn sample_single<R: rand::RngCore + ?Sized>(self, rng: &mut R) -> $target {
+                let low = match self.low {
+                    Bound::Unbounded => panic!("cannot sample {} range unbounded on the left", stringify!($target)),
```

Comment:
Why not `$target::MIN()` for Unbounded on the left?

Reply:
Because that is not unbounded, that’s bounded inclusive. The `rand` API this wraps around follows the same logic.